sensors/repository_sensor.py

```diff
@@ -0,0 +1,221 @@
+            'type': event_type,
+            'payload': payload,
+        }
+        self._increment_event_id()
+
+        self._sensor_service.dispatch(trigger=self._trigger_ref, payload=data)
+
+    # Returns unique id for the dispatching events
+    def _get_event_id(self):
+        ret_id = self._sensor_service.get_value(name=self._trigger_ref)
+        if not ret_id:
+            return 1
+        return ret_id
+
+    # Increments event id of datastore
+    def _increment_event_id(self):
+        self._sensor_service.set_value(name=self._trigger_ref,
+                                       value=int(self._get_event_id()) + 1)
+
+    # initialize last commit for each branches
+    def _init_server_last_commit(self):
+        def _last_ctime(project, repository, branch):
+            _generator = self.client.projects[project].repos[repository].commits(branch['id'])
+            last_commit = [x for x in _generator if x['id'] == branch['latestCommit']]
+
+            if last_commit:
+                return datetime.fromtimestamp(last_commit[0]['authorTimestamp'] / 1000)
+            else:
+                return datetime.strptime('1900-01-01 00:00:00', self.DATE_FORMAT)
+
+        [[self._set_last_commit_time('%s/%s' % (p, r), b['displayId'], _last_ctime(p, r, b))
+            for b in self.client.projects[p].repos[r].branches()]
+            for (p, r) in [x.split('/') for x in self.repositories]]
+
+    def _init_cloud_last_commit(self):
+        _TZ_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+        [[self._set_last_commit_time('%s/%s' % (o, r),
+                                     b.name,
+                                     datetime.strptime(b.commits().next().date, _TZ_FORMAT))
+            for b in Branch.find_branches_in_repository(repository_name=r, client=self.client,
+                                                        owner=o) if isinstance(b, Branch)]
+            for (o, r) in [x.split('/') for x in self.repositories]]
+
+    def _set_last_commit_time(self, repo, branch, last_commit_time):
+        if repo not in self.last_commit:
+            self.last_commit[repo] = {}
+        self.last_commit[repo][branch] = last_commit_time
+
+    def _get_server_branches(self, proj, repo):
+        return [b['displayId'] for b in self.client.projects[proj].repos[repo].branches()]
+
+    def _get_cloud_branches(self, owner, repo):
+        return [b.name for b in Branch.find_branches_in_repository(repository_name=repo,
+                                                                   client=self.client,
+                                                                   owner=owner)]
+
+    def _update_last_commit(self):
+        for commit in self.new_commits:
+            commit_time = datetime.strptime(commit['time'], self.DATE_FORMAT)
+
+            if self.last_commit[commit['repository']][commit['branch']] < commit_time:
+                self.last_commit[commit['repository']][commit['branch']] = commit_time
```

Comment:
To prevent same commits to be replayed when sensor restarts you should also store this value in datastore after each `poll` loop and retrieve it in setup on initialization when server starts up.

You can find examples on how we do that in some other sensors (e.g. Twitter).